src/coord/src/coord.rs

```diff
@@ -867,6 +903,93 @@ impl<S: Append + 'static> Coordinator<S> {
+        if self.pending_writes.is_empty() {
+            return;
+        }
+
+        let timestamp = if let Some(timestamp) = timestamp {
+            timestamp
+        } else {
+            self.peek_local_ts()
+        };
+        let now = (self.catalog.config().now)();
+
+        if timestamp > now {
+            // Cap retry time to 1s. In cases where the system clock has retreated by
+            // some large amount of time, this prevents against then waiting for that
+            // large amount of time in case the system clock then advances back to near
+            // what it was.
+            let remaining_ms =
+                std::cmp::min(timestamp.saturating_sub(now).saturating_add(1), 1_000);
+            let internal_cmd_tx = self.internal_cmd_tx.clone();
+            task::spawn(|| "group_commit", async move {
+                tokio::time::sleep(Duration::from_millis(remaining_ms)).await;
+                internal_cmd_tx
+                    .send(Message::GroupCommit(Some(timestamp)))
+                    .expect("sending to internal_cmd_tx cannot fail");
+            });
+        } else {
+            self.group_commit().await;
+        }
+    }
+
+    /// Commits all pending write transactions at the same timestamp. All pending writes will be
+    /// combined into a single Append command and sent to STORAGE as a single batch. All writes will
+    /// happen at the same timestamp and all involved tables will be advanced to some timestamp
+    /// larger than the timestamp of the write.
+    ///
+    /// If the table that some write was targeting has been deleted while the write was waiting,
+    /// then the write will be ignored and we respond to the client that the write was successful.
+    /// This is only possible if the write and the delete were concurrent. Therefore, we are free to
+    /// order the write before the delete without violating any consistency guarantees.
+    async fn group_commit(&mut self) {
+        // We cannot use the `peek_local_ts` from `try_group_commit`, because it may be invalid.
+        // While group commit was waiting, some DDL may have executed at the peeked timestamp
+        // causing that timestamp to close and become invalid for future writes. Therefore, we must
+        // get a new valid timestamp. This timestamp may be larger than the peeked timestamp and
+        // may even be larger than `now` depending on how much DDL was executed. In the future we
+        // should include DDL in group commits to avoid this issue.
+        let WriteTimestamp {
+            timestamp,
+            advance_to,
+        } = self.get_and_step_local_write_ts();
+        let mut appends: HashMap<GlobalId, Vec<Update<Timestamp>>> = HashMap::new();
+        for PendingWriteTxn { writes, .. } in &mut self.pending_writes {
+            let writes = std::mem::take(writes);
+            for WriteOp { id, rows } in writes {
+                // object may have been deleted while write was waiting
```

Comment:
Because this involves transaction correctness, probably worth duplicating or re-describing the comment above about being able to order concurrent operations however we choose, which is safe under strict serializability.

Reply:
I didn't want to duplicate the comment so I just moved it down here, is that alright? Or do you think it's worth also having it in the doc comments?

Reply:
It's fine just here.